### ISSUE TITLE
Sage combinat widgets & voila

### ISSUE THREAD

**Issue by zerline** [voila-dashboards/voila#607]
When trying to use [_sage_combinat_widgets_](https://github.com/sagemath/sage-combinat-widgets) on _voila_, the following message appears on the browser console:

> The resource from “http://localhost:8888/voila/sage-combinat-widgets.js” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff).

although there is no such file in sage_combinat_widgets code ..

**Comment by zerline** [comment]
**Steps to reproduce**

1.  Dockerfile
FROM sagemath/sagemath:9.1.rc5
RUN sage -pip install --upgrade jupyter_core jupyter_server nbformat nbclient nbconvert pygments voila
RUN sage -pip install --upgrade ipywidgets sage_combinat_widgets
ENTRYPOINT /usr/bin/sage -n jupyter --ip=0.0.0.0 --port=8888 

2. docker build, and docker run with `-ti -p8888:8888`

3. Open a Sagemath notebook, and type:

```
from sage_combinat_widgets import *
GridViewWidget(Partition([4,3]))
```

4. Try to `voila`

**Comment by jtpio** [comment]
Thanks @zerline for the details.

Will try to reproduce that locally. It should also be possible to create a Binder (for example as a [gist](https://gist.github.com)) using the content of the `Dockerfile` posted above, so it can also be tested in the browser.

**Comment by zerline** [comment]
> It should also be possible to create a Binder (for example as a [gist](https://gist.github.com)) using the content of the `Dockerfile` posted above, so it can also be tested in the browser.

like this?

https://gist.github.com/zerline/b8cb987cea0361103ab2fc3204b1b6b9

https://mybinder.org/v2/gist/zerline/b8cb987cea0361103ab2fc3204b1b6b9/master

(timeout right now)

**Comment by jtpio** [comment]
Yes, thanks for putting that together!

It looks like `sage-combinat-widgets` is not published on npm, which seems to be the cause of the error above.

Instead we can use the [enable_nbextension](https://voila.readthedocs.io/en/latest/using.html?highlight=enable_nbextension#using-third-party-widgets-with-voila) flag for voila.

Trying it locally with the following `Dockerfile`:

```dockerfile
FROM sagemath/sagemath:9.1.rc5
RUN sage -pip install --upgrade jupyter_core jupyter_server nbformat nbclient nbconvert pygments voila
RUN sage -pip install --upgrade ipywidgets sage_combinat_widgets

COPY jupyter_config.json .

EXPOSE 8888

ENTRYPOINT /usr/bin/sage -n jupyter --ip=0.0.0.0 --port=8888
```

With `jupyter_config.json`:

```json
{
    "VoilaConfiguration": {
        "enable_nbextensions": true
    }
}
```

And building the image + starting the container with:

```bash
docker build -t sage-docker-test .

docker run -it --rm -p 8888:8888 sage-docker-test
```

The code snippet above should render fine with Voila:

![image](https://user-images.githubusercontent.com/591645/82443153-4e29dd00-9aa1-11ea-8538-87edaa373644.png)

**Comment by zerline** [comment]
> Yes, thanks for putting that together!
> 
Thanks for your quick feedback!

> It looks like `sage-combinat-widgets` is not published on npm, which seems to be the cause of the error above.
>

NB : actually, there is no such "sage-combinat-widgets" JS module. There is some javascript code, but only as small features on top of ipywidgets. I called the module "singleton_widgets", and I did not publish it in its own. 
 
> Instead we can use the [enable_nbextension](https://voila.readthedocs.io/en/latest/using.html?highlight=enable_nbextension#using-third-party-widgets-with-voila) flag for voila.
> 
> Trying it locally with the following `Dockerfile`:
> 
Thanks!

/.../
> The code snippet above should render fine with Voila:
> 

yes, this image is the one expected

**Comment by jtpio** [comment]
Nice!

@zerline let me know if that fixes your issue.

**Comment by zerline** [comment]
Yes!

**Comment by zerline** [comment]
Re-opening. I still have same kind of message:
> The resource from “http://localhost:8888/voila/base/js/namespace.js” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff).

The Dockerfile is here:
https://github.com/sagemath/sage-explorer/blob/voila/Dockerfile

You can test it on binder here:
https://mybinder.org/v2/gh/sagemath/sage-explorer/voila?filepath=demo_sage_explorer.ipynb

and the code to test -- on a Sagemath notebook -- is:
```
from sage_explorer import *
p = Partition([3,3,2,1])
explore(p.cell_poset())
```

**Comment by jtpio** [comment]
Although it takes a little while to load, it seems to be fine on Binder?

![sage-voila](https://user-images.githubusercontent.com/591645/82697586-7cf2af80-9c69-11ea-802e-8b13a69ee5c9.gif)

**Comment by zerline** [comment]
Did you try on the notebook? I get a blank page. Pressing CTRL-shift helps, though ..

BTW on JLab, I can't open a new tab (no '+' button), and I can't find any way to get voilà working, either. Firefox 68.7 or Chromium.

**Comment by jtpio** [comment]
> Did you try on the notebook? I get a blank page. Pressing CTRL-shift helps, though ..

Yes that was after starting a Binder from `master` and creating a new notebook. Maybe worth clearing the browser cache just to be sure?


> BTW on JLab, I can't open a new tab (no '+' button), and I can't find any way to get voilà working, either. Firefox 68.7 or Chromium 80.0.3987.162.

Is the preview extension installed? (`jupyter labextension install @jupyter-voila/jupyterlab-preview`)

Otherwise would it be possible to post a screenshot?

**Comment by zerline** [comment]
> > Did you try on the notebook? I get a blank page. Pressing CTRL-shift helps, though ..
> 
> Yes that was after starting a Binder from ~`master`~ the `voila` branch and creating a new notebook. 

a classical notebook, from outside JLab?

>Maybe worth clearing the browser cache just to be sure?
> 
> > BTW on JLab, I can't open a new tab (no '+' button), and I can't find any way to get voilà working, either. Firefox 68.7 or Chromium 80.0.3987.162.
> 
> Is the preview extension installed? (`jupyter labextension install @jupyter-voila/jupyterlab-preview`)
> 
I cant' find it. Does it depend on JLab *2* ?
 
> Otherwise would it be possible to post a screenshot?

**Comment by jtpio** [comment]
> a classical notebook, from outside JLab?

Yes, with the "New" dropdown menu and selecting the SageMath kernel.

> I cant' find it. Does it depend on JLab _2_ ?

It would have to be added to the `Dockerfile` so it becomes available on Binder.
